hugr-core/src/envelope/description.rs

```diff
@@ -0,0 +1,567 @@
+    pub fn packaged_extensions(&self) -> impl Iterator<Item = &Option<ExtensionDesc>> {
+        self.packaged_extensions.iter()
+    }
+}
+
+/// High level description of an extension.
+#[derive(derive_more::Display, Debug, Clone, PartialEq, serde::Deserialize, serde::Serialize)]
+#[display("Extension {name} v{version}")]
+pub struct ExtensionDesc {
+    /// Name of the extension.
+    pub name: String,
+    /// Version of the extension.
+    pub version: Version,
+}
+
+impl ExtensionDesc {
+    /// Create a new extension description.
+    pub fn new(name: impl ToString, version: impl Into<Version>) -> Self {
+        Self {
+            name: name.to_string(),
+            version: version.into(),
+        }
+    }
+}
+
+impl<E: AsRef<crate::Extension>> From<&E> for ExtensionDesc {
+    fn from(ext: &E) -> Self {
+        let ext = ext.as_ref();
+        Self {
+            name: ext.name.to_string(),
+            version: ext.version.clone(),
+        }
+    }
+}
+
+fn extend_option_vec<T: Clone>(vec: &mut Option<Vec<T>>, items: impl IntoIterator<Item = T>) {
+    if let Some(existing) = vec {
+        existing.extend(items);
+    } else {
+        vec.replace(items.into_iter().collect());
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, serde::Serialize)]
+/// Description of the entrypoint of a module.
+pub struct Entrypoint {
+    /// Node id of the entrypoint.
+    pub node: Node,
+    #[serde(serialize_with = "op_serialize")]
+    /// Operation type of the entrypoint node.
+    pub optype: OpType,
+}
+
+impl Entrypoint {
+    /// Create a new entrypoint description.
+    pub fn new(node: Node, optype: OpType) -> Self {
+        Self { node, optype }
+    }
+}
+
+/// Get a string representation of an OpType for description purposes.
+pub fn op_string(op: &OpType) -> String {
+    match op {
+        OpType::FuncDefn(defn) => format!(
+            "FuncDefn({})",
+            func_symbol(defn.func_name(), defn.signature())
+        ),
+        OpType::FuncDecl(decl) => format!(
+            "FuncDecl({})",
+            func_symbol(decl.func_name(), decl.signature())
+        ),
+        OpType::DFG(dfg) => format!("DFG({})", dfg.signature()),
+        _ => format!("{op}"),
+    }
+}
+fn op_serialize<S>(op_type: &OpType, serializer: S) -> Result<S::Ok, S::Error>
+where
+    S: serde::Serializer,
+{
+    serializer.serialize_str(op_string(op_type).as_str())
+}
+
+#[derive(Debug, Clone, PartialEq, Default, serde::Serialize)]
+/// High-level description of a module in a HUGR package.
+pub struct ModuleDesc {
+    /// Number of nodes in the module.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub num_nodes: Option<usize>,
+    /// The entrypoint node and the corresponding operation type.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub entrypoint: Option<Entrypoint>,
+    /// Extensions used in the module computed while resolving, expected to be a subset of `used_extensions_metadata`.
```

Comment:
```suggestion
    /// Extensions used in the module computed while resolving, expected to be a subset of `used_extensions_generator`.
```
?
The `_metadata` is only used in local variable names.